### ISSUE TITLE
GetAllStrings throws "No manifests exist for the current culture"

### ISSUE THREAD

**Issue by aspnet-hello** [aspnet/Home#2630]
_From @atuzovic on Sunday, December 24, 2017 7:56:21 AM_

We have updated all our solution assemblies from 1.1.x to 2.0.1 and out of sudden this message has started to appear. 

We have an MVC project that references assembly AppImpact.AspNetCore.

In assembly A we have resources for two cultures as shown on the image:
- en
- bs-Latn-BA

![image](https://user-images.githubusercontent.com/5385262/34327624-87ea7fc2-e8c9-11e7-949e-778b52b25c97.png)

In MVC project, we create a localizer by using Create method on IStringLocalizerFactory with the following parameters (as shown on the images):
- "Buttons" as baseName
- "AppImpact.AspNetCore" as location

![image](https://user-images.githubusercontent.com/5385262/34327639-cea9715c-e8c9-11e7-8272-8b84ce60d608.png)

![image](https://user-images.githubusercontent.com/5385262/34327644-e58b22bc-e8c9-11e7-9d18-ea8b26d14d17.png)

If we inspect created localizer instance, we can see that assembly is present in _resourceNamesCache:

![image](https://user-images.githubusercontent.com/5385262/34327661-5a411a08-e8ca-11e7-97e3-391ab9e4eb80.png)

Highlighted value is:
{[Assembly=AppImpact.AspNetCore, Version=1.2.2.0, Culture=bs-Latn-BA, PublicKeyToken=null;resourceName=AppImpact.AspNetCore.Resources.Buttons, ]}

As soon as we try to iterate through results of GetAllStrings method, we get "No manifests exist for the current culture".

![image](https://user-images.githubusercontent.com/5385262/34327669-d22d3498-e8ca-11e7-8bd9-820c362a6d2f.png)

This code was in production for more than a year. Only difference that was done was upgrade to 2.x.

Funny thing is that as soon as we rename Buttons.en.resx (or any other for that matter) to Buttons.resx (old logic with default resource), everything starts to work.

We are looking for a better solution rather than doing a mass rename of resource files.

Is this behavior as expected or there is a bug?

EDIT:

We have tried now the second overload of GetAllStrings method:

![image](https://user-images.githubusercontent.com/5385262/34327782-bcfcc478-e8cd-11e7-9467-072bcd8041d3.png)

And now we get a different error:

![image](https://user-images.githubusercontent.com/5385262/34327785-d298f4fa-e8cd-11e7-9132-3bf5a11c2cf9.png)

However, in the bin folder where the build is, there is a resource file for this assembly:

![image](https://user-images.githubusercontent.com/5385262/34327788-fdf364be-e8cd-11e7-9a79-a05efd7c2611.png)

_Copied from original issue: aspnet/Localization#433_

**Comment by aspnet-hello** [comment]
_From @hishamco on Sunday, December 24, 2017 9:05:11 AM_

This is remind me of an old bug #257, but it was solved by a temporary fix. Could you please create a minimal repo to reproduce the issue?

**Comment by aspnet-hello** [comment]
_From @atuzovic on Sunday, December 24, 2017 11:36:20 AM_

Thank you for your quick response. :)

https://github.com/App-Impact-Labs/ISSUE-MissingResourceManifestExample

1. Run Xampl.Services
2. Navigate to /StaticResources/Get

You will get exception immediately.

**Comment by aspnet-hello** [comment]
_From @hishamco on Sunday, December 24, 2017 11:44:23 AM_

Let me check ..

**Comment by aspnet-hello** [comment]
_From @hishamco on Sunday, December 24, 2017 12:18:29 PM_

Removing the `en` from `Buttons.en.resx` fixing the issue for me, because the `en` is not needed in case that English is the default culture.

The `GetAllStrings(false)` is not useful in your case while ther's no parent cultures provided

Also you can use `app.UseRequestLocalization` instead of setting Bosnia and Herzegovina the culture in the controller

**Comment by aspnet-hello** [comment]
_From @atuzovic on Sunday, December 24, 2017 12:26:41 PM_

RequestLocalization will just set CultureInfo based on inputs (query string, cookie, header).

In production we are using it, but I have removed it from here for simplicity reasons.

Yes, removing "en" culture tag from resource file name does the thing, but that is "dirty" fix as general recommendation is not to have default RESX file.

Here is one more thing that you should try - now that you have removed "en" prefix from default resource, it will work (you will not get an exception), but check what values are you getting from GetAllStrings.

You should get translations for Bosnian language - but that's not the case, so even with removal of "en" prefix, it is still not working properly.

Again, this was all working as intended with 1.x versions. It is 2.x versions that have caused for this to happen.

**Comment by aspnet-hello** [comment]
_From @hishamco on Sunday, December 24, 2017 12:32:54 PM_

> Yes, removing en from resources does the thing, but that is "dirty" fix as general recommendation is not to have default RESX file.

I don't think it's a dirty fix 😄, this the default for any resource, while the default culture in .NET is en-US, I will check if this documented in the docs or I will suggest it

Yep I got the Bosnian translation, which is the right values if I'm not wrong
> You should get translations for Bosnian language - but that's not the case, so even with removal of "en" prefix, it is still not working properly.

So what's the result that you expect?!!

**Comment by aspnet-hello** [comment]
_From @atuzovic on Sunday, December 24, 2017 12:48:05 PM_

From official documentation (https://docs.microsoft.com/en-us/aspnet/core/fundamentals/localization):

![image](https://user-images.githubusercontent.com/5385262/34329077-d2432d5c-e8f2-11e7-8227-b26d32bde3ef.png)

Problem is that we have several solutions in the production that have followed this guideline.

What you are now suggesting will force us to do rename of all resource files and to also change a lot of our code that was built with presumption that there is no default resource file, but instead that there is a default culture (why otherwise set DefaultRequestCulture in RequestLocalizationMiddleware).

As for the second part - can you do a pull now, build and run again and see if you get results on Bosnian?

I have added Buttons.resx but left Buttons.en.resx as well.

**Comment by aspnet-hello** [comment]
_From @hishamco on Sunday, December 24, 2017 1:00:40 PM_

I read it before a minute ago, nothing but this happen behind this by setting the `NeutralResourcesLanguage` attribute, so if use the following in your `AssemblyInfo.cs`:
```
[assembly: NeutralResourcesLanguage("bs-Latn-BA")]
```
the `Buttons.resx` will be the default resource, so if it's contains the Bosinian resources it will load them.
That's why you see this behavior, i'm not forcing you 😄, and with the `DefaultRequestCulture` you can set your default culture which is useful if the current culture is not supported the localization APIs will use it as default one

**Comment by aspnet-hello** [comment]
_From @atuzovic on Sunday, December 24, 2017 1:09:59 PM_

We don't have AssemblyInfo.cs at all, but regardless of that, some libraries here are part of our internal framework that is used across different solutions, and almost each solution has a different default culture. 

We cannot set default culture to be bs-Latn-BA on internal framework libraries, as it is not guaranteed it is going to be default on any solutions that depend on it. 

Again, why did this stop to work with 2.x while it was working as expected with 1.x versions? What has been changed? We are trying to find some explanation over this, but nothing is available online.

**Comment by aspnet-hello** [comment]
_From @hishamco on Sunday, December 24, 2017 1:16:56 PM_

I knew you don't have `AssemblyInfo.cs` but I would like to explain to you why this happen
> What has been changed?

AFAIK, the big once are:

- Remove dependency on Microsoft.AspNetCore.Hosting.Abstractions
- Consider surfacing additional diagnostics information as part of LocalizedString when a resource is not found

for more details you can check the [releases](https://github.com/aspnet/Localization/releases)

**Comment by aspnet-hello** [comment]
_From @atuzovic on Sunday, December 24, 2017 1:37:12 PM_

Thank you really for your efforts, but so far it was not really helpful.

There has been a change in internal implementation and that change is making a huge problem for us at the moment. I am hoping that someone from the development team will shed some light on this.

Apart from the issue that we are facing, fact that adding .resx and .en.resx at the same time disables reading of existing .bs-Latn-BA.resx does not seem like everything is working as it should.

**Comment by aspnet-hello** [comment]
_From @hishamco on Sunday, December 24, 2017 1:42:44 PM_

@atuzovic just one more question, can you let me know why you adding both `.resx` and `.en.resx` at the same time?!!

**Comment by aspnet-hello** [comment]
_From @atuzovic on Sunday, December 24, 2017 1:47:00 PM_

Just for fun - to see the behavior. I don't see valid explanation why is it disabling .bs-Latn-BA.resx resource.

But again, that's not the point.

Point is - our solution(s) have stopped working after the upgrade and we have documented what is happening.

I am asking for official clarification/recommendation on upgrade path from 1.x to 2.x.

Can you please let us know that this is all working as expected, there is no bug, and that we should rewrite a bunch of our code / rename tons of resource files to be able to run our code?

**Comment by aspnet-hello** [comment]
_From @hishamco on Sunday, December 24, 2017 1:55:29 PM_

Seems having both `.resx` and `.en.resx` causes a bug, perhaps this is come from underlying `ResourceManager` APIs, @ryanbrandenburg @pranavkm any idea about this?

BTW @atuzovic there's a localization issue that I knew in 2.0.0, could you please try to upgrade to the latest preview bits and have a try. If it's still the same hope one of the team will respond

**Comment by aspnet-hello** [comment]
_From @atuzovic on Sunday, December 24, 2017 2:00:06 PM_

Update to latest preview bits of what?

Solution is currently using 2.0.1 version of Microsoft.Extension.Localization. I don't see any preview available on NuGet:
https://www.nuget.org/packages/Microsoft.Extensions.Localization/

Do you want me to add specific branch of this repository to solution directly or you're suggesting something else?

**Comment by aspnet-hello** [comment]
_From @hishamco on Sunday, December 24, 2017 2:01:52 PM_

I suggest to use the dev feed (nightly builds), look i will do it in your behalf and waiting until one of the team respond ..

**Comment by aspnet-hello** [comment]
_From @atuzovic on Sunday, December 24, 2017 2:12:21 PM_

Alright, I have updated solution to nightly build (2.1.0-preview1-27907) and it is working the same way it was working with 1.x versions. I have removed Buttons.resx from solution => there is no error and I get proper translations for bs-Latn-BA.

I have made additional commit, so you can pull solution yourself and give it a try.

So 2.0.1 has a bug clearly. Do you have any ETA on release of 2.1.0 ?

**Comment by aspnet-hello** [comment]
_From @hishamco on Sunday, December 24, 2017 2:20:38 PM_

As I told you before there's a bug I remeber in `IViewLocalizer` which I blogged about it [here](http://www.hishambinateya.com/localization-issue-in-razorpages), but I'm not sure If it's the reason behind this behavior or there's something else had been fixed
> Do you have any ETA on release of 2.1.0 

Nope, but you can checkout https://github.com/aspnet/Home/wiki/Roadmap

**Comment by aspnet-hello** [comment]
_From @atuzovic on Sunday, December 24, 2017 8:39:47 PM_

Q1-Q2 2018 seems a bit undeterministic (probably targeting Build 2018) and we are stuck halfway at migrating from 1.x to 2.x.

What are the changes to get hotfix 2.0.2 for this issue in the meantime?

Again, apart from doing a mass rename, which will also affect csproj files and cause us to rewrite bunch of our code that handles versioning of translations, we don't see an easy fix to survive till 2.1.0.

Also, using nightly builds in production is not an option for us.

**Comment by aspnet-hello** [comment]
_From @hishamco on Sunday, December 24, 2017 8:43:10 PM_

> What are the changes to get hotfix 2.0.2 for this issue in the meantime?

I think there's a patch release which I don't have it's link, but I will look for it
> Also, using nightly builds in production is not an option for us.

Yep using nightly builds in production is a good option

**Comment by aspnet-hello** [comment]
_From @hishamco on Sunday, December 24, 2017 8:51:57 PM_

I found these links [2.0.3](https://github.com/dotnet/core/blob/master/release-notes/download-archives/2.0.3.md) and [2.0.4](https://github.com/dotnet/core/blob/master/release-notes/download-archives/2.0.4-download.md) if you are running on premise

**Comment by aspnet-hello** [comment]
_From @atuzovic on Sunday, December 24, 2017 10:07:37 PM_

If you are referring to .NET Core version, I have 2.0.4. on my dev machine already:

![image](https://user-images.githubusercontent.com/5385262/34333971-27552ee0-e942-11e7-9ae7-291c921397e3.png)

I was asking about release of 2.0.2. for Microsoft.Extensions.Localization library.

**Comment by aspnet-hello** [comment]
_From @hishamco on Sunday, December 24, 2017 11:07:01 PM_

Oops, sorry it's my mistake

**Comment by aspnet-hello** [comment]
_From @mkArtakMSFT on Wednesday, December 27, 2017 9:27:12 AM_

@ryanbrandenburg, can you please check and see which change fixed the issue in 2.1.0 release (from 2.0) and see whether we can come up with a simpler workaround?

**Comment by aspnet-hello** [comment]
_From @MookieFumi on Saturday, December 30, 2017 2:21:55 AM_

The same issue like another mates. 

```csharp
var strings = _sharedLocalizer.GetAllStrings();
```

**Message: No manifests exist for the current culture.**

StackTrace:
```
 at Microsoft.Extensions.Localization.ResourceManagerStringLocalizer.GetResourceNamesFromCultureHierarchy(CultureInfo startingCulture)
   at Microsoft.Extensions.Localization.ResourceManagerStringLocalizer.<GetAllStrings>d__15.MoveNext()
   at System.Collections.Generic.LargeArrayBuilder`1.AddRange(IEnumerable`1 items)
   at System.Collections.Generic.EnumerableHelpers.ToArray[T](IEnumerable`1 source)
   at System.Linq.Enumerable.ToArray[TSource](IEnumerable`1 source)
   at System.Linq.SystemCore_EnumerableDebugView`1.get_Items()
```

```
> dotnet --info                                       
Herramientas de la línea de comandos .NET (2.1.3)     
                                                      
Product Information:                                  
 Version:            2.1.3                            
 Commit SHA-1 hash:  a0ca411ca5                       
                                                      
Runtime Environment:                                  
 OS Name:     Windows                                 
 OS Version:  10.0.16299                              
 OS Platform: Windows                                 
 RID:         win10-x64                               
 Base Path:   C:\Program Files\dotnet\sdk\2.1.3\      
                                                      
Microsoft .NET Core Shared Framework Host             
                                                      
  Version  : 2.0.4                                    
  Build    : 7f262f453d8c8479b9af91d34c013b3aa05bc1ff 
```

Forwarded from https://github.com/aspnet/Localization/issues/406
Thank you so much.

**Comment by ryanbrandenburg** [comment]
The change that fixed this for 2.1.0 was aspnet/Localization#413. The original problem was that corefx removed the GetResourceSet API from ResourceManager, we worked around that with aspnet/Localization#264 (essentially trying to get the full resource set ourselves by hand) but that solution had the problem we see here which is why we started using the GetResourceSet API again once it was available.

After experimenting for a bit I couldn't find any way to "work around" this while still using ResourceManagerStringLocalizer in 2.0.x. If we think this meets the bar for patch I believe I know the general fix to make. Experience in this area tells me there will be edge cases which our improved solution still doesn't quite work for because we don't have the full context and platform knowledge that GetResourceSet does.

**Comment by mkArtakMSFT** [comment]
@ryanbrandenburg, can you please investigate this further and see what options do we have? Is this really something customers can confidently work around or not?

**Comment by ryanbrandenburg** [comment]
I can confirm that the workaround of having the resource for your neutral language be named `Test.resx` instead of `Test.en-US.resx` works.

**Comment by hishamco** [comment]
I quote from what @atuzovic said before
> adding `.resx` and `.en.resx` at the same time disables reading of existing `.bs-Latn-BA.resx` does not seem like everything is working as it should

**Comment by amoerie** [comment]
This is a workaround that is confirmed working for us, until this issue is resolved:
``` csharp
    private static readonly FieldInfo ResourceManagerField = typeof(ResourceManagerStringLocalizer)
      .GetField("_resourceManager", BindingFlags.Instance | BindingFlags.NonPublic);

    public IDictionary<string, string> GetAllStrings(IStringLocalizer stringLocalizer) {
        if (!(stringLocalizer is ResourceManagerStringLocalizer localizer)) 
            throw new ArgumentException("Until https://github.com/aspnet/Home/issues/2630 is resolved, a ResourceManagerStringLocalizer is required");

        if (!(ResourceManagerField.GetValue(localizer) is ResourceManager resourceManager))
            throw new ArgumentException("Expected a ResourceManager here");
        
	var resourceSet = resourceManager.GetResourceSet(CultureInfo.CurrentUICulture, true, false);
        if (resourceSet == null) ...; // do something meaningful. This can happen when you create an IStringLocalizer for an MVC controller which does not have a corresponding resx file.
        
	return resourceSet.Cast<DictionaryEntry>()
          .ToDictionary(entry => entry.Key.ToString(), entry => entry.Value.ToString());
    }
```

**Comment by ryanbrandenburg** [comment]
In response to this we've decided that we also need to clean up and clarify our localization documentation in this area, so I've filed aspnet/Home#2905.

**Comment by Eilon** [comment]
Just checking: we can consider this for the next patch, but was the work already done for 2.1 (dev)? Or is there a separate issue tracking tha?

**Comment by ryanbrandenburg** [comment]
Yes, this work is done for 2.1. We fixed it in aspnet/Localization#413 when they added the GetAllStrings API back to ResourceManager.

**Comment by Eilon** [comment]
Discussed within the team and we feel that given there is a workaround for this issue in 2.0, and that it's fixed in the upcoming 2.1 release, we will not patch this.
